src/Common/src/System/Collections/Generic/EnumerableHelpers.Linq.cs

```diff
@@ -0,0 +1,46 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Collections.Generic
+{
+    /// <summary>Internal helper functions for working with enumerables.</summary>
+    internal static partial class EnumerableHelpers
+    {
+        /// <summary>Gets the count of the specified enumerable.</summary>
+        /// <param name="source">The enumerable to count.</param>
+        /// <returns>The count of the enumerable.</returns>
+        internal static int Count<T>(IEnumerable<T> source)
+        {
+            Debug.Assert(source != null);
+
+            var collectionOfT = source as ICollection<T>;
+            if (collectionOfT != null)
+            {
+                return collectionOfT.Count;
+            }
+            
+            var collection = source as ICollection;
+            if (collection != null)
+            {
+                return collection.Count;
+            }
+            
+            int count = 0;
+            using (IEnumerator<T> e = source.GetEnumerator())
```

Comment:
@stephentoub Ok. I was trying to say that if we cared about optimizing this function at all, we could have just written the whole thing inline like I did [here](https://github.com/jamesqo/corefx/blob/4e53c20c25299766037fc076446b426237106223/src/System.Linq/src/System/Linq/Where.cs#L435-L457) and avoid any typecasting. But since we are running the selector anyway, I don't think it's worth trying to optimize these functions; we should not add any additional complexity trying to do so, even if that is just one additional method.

I think this can be revisited in another PR if it turns out to be an issue.